app/scripts/controllers/preferences.js

```diff
@@ -475,20 +480,59 @@ export default class PreferencesController {
   /**
    * updates custom RPC details
    *
-   * @param {string} url - The RPC url to add to frequentRpcList.
-   * @param {string} chainId - Optional chainId of the selected network.
-   * @param {string} ticker   - Optional ticker symbol of the selected network.
-   * @param {string} nickname - Optional nickname of the selected network.
+   * @param {string} rpcUrl - The RPC url to add to frequentRpcList.
+   * @param {string} chainId - The chainId of the selected network.
+   * @param {string} [ticker] - Optional ticker symbol of the selected network.
+   * @param {string} [nickname] - Optional nickname of the selected network.
+   * @param {object} [rpcPrefs] - Optional RPC preferences, such as the block explorer URL
    *
    */
-  updateRpc (newRpcDetails) {
+  async updateRpc (newRpcDetails) {
     const rpcList = this.getFrequentRpcListDetail()
     const index = rpcList.findIndex((element) => {
       return element.rpcUrl === newRpcDetails.rpcUrl
     })
     if (index > -1) {
       const rpcDetail = rpcList[index]
       const updatedRpc = { ...rpcDetail, ...newRpcDetails }
+      if (rpcDetail.chainId !== updatedRpc.chainId) {
+
+        // When the chainId is changed, associated address book entries should
+        // also be migrated. The address book entries are keyed by the `network` state,
+        // which for custom networks is the chainId with a fallback to the networkId
+        // if the chainId is not set.
+
+        let addressBookKey = rpcDetail.chainId
+        if (!addressBookKey) {
+          // We need to find the networkId to determine what these addresses were keyed by
+          const provider = new ethers.providers.JsonRpcProvider(rpcDetail.rpcUrl)
+          try {
+            addressBookKey = await provider.send('net_version')
+            assert(typeof addressBookKey === 'string')
+          } catch (error) {
+            log.debug(error)
+            log.warn(`Failed to get networkId from ${rpcDetail.rpcUrl}; skipping address book migration`)
+          }
+        }
+
+        // There is an edge case where two separate RPC endpoints are keyed by the same
+        // value. In this case, the contact book entries are duplicated so that they remain
+        // on both networks, since we don't know which network each contact is intended for.
+
+        let duplicate = false
+        const buildInProviderNetworkIds = Object.values(NETWORK_TYPE_TO_ID_MAP)
+          .map((ids) => ids.networkId)
```

Comment:
Do we want to use `chainId` instead of `networkId` here, or is this because we're comparing at legacy values, which were assigned by `networkId`?

Reply:
It's the latter; we're looking for the value that was used as the key prior `chainId` being made mandatory for custom RPC endpoints.